File: Data Structures & Algorithms/two-integer-sum/submission-0.py
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int] | None:
        if not nums: return None

        seen = dict()
        for i in range(len(nums)):
            if nums[i] in seen:
                return [seen[nums[i]], i]
            seen[target-nums[i]] = i
        return None

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
#include <vector>

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        std::vector<int> prefix(n,1);
        std::vector<int> suffix(n,1);
        for(int i=1;i<n;i++) {
            prefix[i]=prefix[i-1]*nums[i-1];
        }
        for(int i=1;i<n;i++){
            suffix[n-1-i] = suffix[n-i]*nums[n-i];
        }
        std::vector<int> out(n);
        for(int i = 0;i<n;i++) {
            out[i] = prefix[i]*suffix[i];
        }
        return out;
    }
};
